src/v/cloud_topics/level_one/metastore.h

```diff
@@ -109,6 +110,96 @@ class metastore {
     // `out_of_range`.
     virtual ss::future<std::expected<object_response, errc>>
     get_first_ge(const model::topic_id_partition&, model::timestamp) = 0;
+
+    // Compaction metadata updates per partition
+    //
+    // Kafka compaction works by taking "dirty" ranges of data, collecting the
+    // keys and offsets within that range, and then removing all older
+    // instances of those keys from the beginning of the log. At that point,
+    // that range of data is considered "cleaned". For all keys in a cleaned
+    // range, only the latest version of that key exists in the log up to the
+    // cleaned range.
+    //
+    // Kafka also has the concept of a "tombstone" (empty-value record) that
+    // indicates the logical deletion of the record's key. Once a given
+    // tombstone record has been cleaned, a timer begins and after
+    // delete.retention.ms elapses, the cleaned tombstone record may be
+    // removed.
+    //
+    // To support these ideas, the metastore tracks cleaned offset ranges and
+    // whether they contain tombstones. This allows it to expose dirty ranges
+    // and ranges with removable tombstones to callers.
+    struct compaction_update {
+        bool operator==(const compaction_update&) const = default;
+        struct cleaned_range {
+            bool operator==(const cleaned_range&) const = default;
+            kafka::offset base_offset;
+            kafka::offset last_offset;
+            std::optional<model::timestamp> cleaned_with_tombstones_at;
```

Comment:
I'm assuming that if the `std::optional` is set then it means that there is both tombstones and this is the timestamp that compaction finished? I do wonder if from an API prospective, if it's easier to understand if we always pass in a timestamp and a separate bool for if there were tombstones, then the impl just ignores the timestamp otherwise. I'm fine to keep the `std::optional`, but let's add a quick note so folks don't have to guess or go read more code.